sklearn/utils/estimator_checks.py

```diff
@@ -1192,6 +1193,39 @@ def check_array_api_input_and_values(
     )
 
 
+def check_estimator_sparse_tag(name, estimator_orig):
+    if SPARSE_ARRAY_PRESENT:
+        sparse_container = sparse.csr_array
+    else:
+        sparse_container = sparse.csr_matrix
+    estimator = clone(estimator_orig)
+
+    rng = np.random.RandomState(0)
+    n_samples = 15 if name == "SpectralCoclustering" else 40
+    X = rng.uniform(size=(n_samples, 3))
+    X[X < 0.6] = 0
+    y = rng.randint(0, 3, size=n_samples)
+    X = _enforce_estimator_tags_X(estimator, X)
+    y = _enforce_estimator_tags_y(estimator, y)
+    X = sparse_container(X)
+
+    tags = get_tags(estimator)
+    if tags.input_tags.sparse:
+        estimator.fit(X, y)  # should pass
```

Comment:
I'm thinking that we could help third-party library a bit better here by re-raising a better error message.

```suggestion
        try:
            estimator.fit(X, y)  # should pass
        except Exception as e:
            raise AssertionError(
                f"Estimator {name} raised an exception: {e}. The tag "
                "self.input_tags.sparse_tag might not be consistent with the "
                "estimator's ability to handle sparse data (i.e. controlled by the "
                "parameter `accept_sparse` in `validate_data` or `check_array` "
                f"functions). Got input_tags.sparse={tags.input_tags.sparse}."
            )
```